// @ngrx/router-store/esm5/src/router_store_module.ngfactory.js
/**
 * @fileoverview This file was generated by the Angular template compiler. Do not edit.
 *
 * @suppress {suspiciousCode,uselessCode,missingProperties,missingOverride,checkTypes}
 * tslint:disable
 */ 
export { InjectionToken_1 as InjectionToken_14 } from "ngrx/modules/store/src/tokens.ngfactory";
export { Inject_7 as Inject_21, ɵangular_packages_core_core_a_8 as ɵangular_packages_core_core_a_22, Observable_1 as Observable_24, BehaviorSubject_6 as BehaviorSubject_27 } from "ngrx/modules/store/src/reducer_manager.ngfactory";
export { Store as Store_23 } from "@ngrx/store/src/store";
export { StateObservable as StateObservable_25 } from "@ngrx/store/src/state";
export { ActionsSubject as ActionsSubject_26 } from "@ngrx/store/src/actions_subject";
export { ReducerManager as ReducerManager_28, ReducerManagerDispatcher as ReducerManagerDispatcher_32 } from "@ngrx/store/src/reducer_manager";
export { INITIAL_STATE as INITIAL_STATE_29, INITIAL_REDUCERS as INITIAL_REDUCERS_30, REDUCER_FACTORY as REDUCER_FACTORY_31 } from "@ngrx/store/src/tokens";
export { Router as Router_35 } from "@angular/router";
import * as i0 from "@angular/core";
import * as i1 from "ngrx/modules/router-store/src/router_store_module";
import * as i2 from "ngrx/modules/router-store/src/serializer";
import * as i3 from "@ngrx/store/src/store";
import * as i4 from "@angular/router";
var StoreRouterConnectingModuleNgFactory = i0.ɵcmf(i1.StoreRouterConnectingModule, [], function (_l) { return i0.ɵmod([i0.ɵmpd(512, i0.ComponentFactoryResolver, i0.ɵCodegenComponentFactoryResolver, [[8, []], [3, i0.ComponentFactoryResolver], i0.NgModuleRef]), i0.ɵmpd(512, i2.RouterStateSerializer, i2.DefaultRouterStateSerializer, []), i0.ɵmpd(256, i1._ROUTER_CONFIG, i1.ɵ0, []), i0.ɵmpd(1024, i1.ROUTER_CONFIG, i1._createDefaultRouterConfig, [i1._ROUTER_CONFIG]), i0.ɵmpd(1073742336, i1.StoreRouterConnectingModule, i1.StoreRouterConnectingModule, [i3.Store, i4.Router, i2.RouterStateSerializer, i1.ROUTER_CONFIG])]); });
export { StoreRouterConnectingModuleNgFactory as StoreRouterConnectingModuleNgFactory };
