// @ngrx/store-devtools/esm5/src/instrument.ngfactory.js
/**
 * @fileoverview This file was generated by the Angular template compiler. Do not edit.
 *
 * @suppress {suspiciousCode,uselessCode,missingProperties,missingOverride,checkTypes}
 * tslint:disable
 */ 
export { InjectionToken_1 as InjectionToken_1 } from "ngrx/modules/store/src/tokens.ngfactory";
export { StateObservable as StateObservable_15 } from "@ngrx/store/src/state";
export { Observable_1 as Observable_16, BehaviorSubject_3 as BehaviorSubject_19 } from "ngrx/modules/store/src/state.ngfactory";
export { ReducerManagerDispatcher as ReducerManagerDispatcher_17 } from "@ngrx/store/src/reducer_manager";
export { ActionsSubject as ActionsSubject_18 } from "@ngrx/store/src/actions_subject";
import * as i0 from "@angular/core";
import * as i1 from "ngrx/modules/store-devtools/src/instrument";
var StoreDevtoolsModuleNgFactory = i0.ɵcmf(i1.StoreDevtoolsModule, [], function (_l) { return i0.ɵmod([i0.ɵmpd(512, i0.ComponentFactoryResolver, i0.ɵCodegenComponentFactoryResolver, [[8, []], [3, i0.ComponentFactoryResolver], i0.NgModuleRef]), i0.ɵmpd(1073742336, i1.StoreDevtoolsModule, i1.StoreDevtoolsModule, [])]); });
export { StoreDevtoolsModuleNgFactory as StoreDevtoolsModuleNgFactory };
